File: todo-api-rs/src/application/use_cases/todo/update.rs
use crate::application::dtos::todo::update::{UpdateTodoError, UpdateTodoInput, UpdateTodoOutput};
use crate::application::repositories::todo::{TodoRepository, UpdateError, UpdateQuery};
use crate::domain::use_case::UseCase;

#[derive(Debug)]
pub struct UpdateTodoUseCase<T> {
    repository: T,
}

impl<T: TodoRepository> UpdateTodoUseCase<T> {
    pub fn new(repository: T) -> Self {
        Self { repository }
    }
}

impl<T: TodoRepository> UseCase<UpdateTodoInput, UpdateTodoOutput> for UpdateTodoUseCase<T> {
    async fn exec(mut self, input: UpdateTodoInput) -> UpdateTodoOutput {
        let query = UpdateQuery {
            id: input.id,
            title: input.title.clone(),
            description: input.description,
            status: input.status,
            todo_at: input.todo_at,
        };

        self.repository
            .update(query)
            .await
            .map_err(|err| match err {
                UpdateError::NotFound => UpdateTodoError::NotFound,
                UpdateError::DuplicatedTitle => UpdateTodoError::DuplicatedTitle(input.title),
                UpdateError::Internal(err) => UpdateTodoError::Internal(err),
            })
    }
}

File: todo-api-js/src/application/use-cases/todo/update.ts
import { InternalError } from "@domain/utils/exception";
import { Err } from "@domain/utils/result";
import { type UseCase } from "@domain/use-case";
import { type UpdateTodoInput, type UpdateTodoOutput } from "@application/dtos/todo/update";
import { type TodoRepository } from "@application/repositories/todo";

export class UpdateTodoUseCase implements UseCase<UpdateTodoInput, UpdateTodoOutput> {
    constructor(private repository: TodoRepository) {}

    async exec(input: UpdateTodoInput): Promise<UpdateTodoOutput> {
        try {
            return this.repository.update(input);
        } catch (e) {
            const error = new InternalError(e instanceof Error ? e : undefined);
            return new Err(error);
        }
    }
}
